File: 1520-maximum-number-of-non-overlapping-substrings/1520-maximum-number-of-non-overlapping-substrings.java
class Solution {
    public List<String> maxNumOfSubstrings(String s) {
        int[] count = new int[26];
        int[] first = new int[26];
        int[] last = new int[26];

        Arrays.fill(first, -1);
        Arrays.fill(last, -1);

        List<Integer> order = new ArrayList<>();

        for (int i = 0; i < s.length(); i++) {
            int c = s.charAt(i) - 'a';

            if (count[c] == 0) {
                first[c] = i;
                order.add(c);
            }

            count[c]++;
            last[c] = i;
        }

        List<String> res = new ArrayList<>();
        Deque<int[]> queue = new ArrayDeque<>();

        for (int c : order) {
            queue.addFirst(new int[]{
                first[c], last[c], count[c]
            });

            int left = Integer.MAX_VALUE;
            int right = Integer.MIN_VALUE;
            int total = 0;

            for (int[] item : queue) {
                total += item[2];
                left = Math.min(left, item[0]);
                right = Math.max(right, item[1]);

                if (total == right - left + 1) {
                    break;
                }
            }

            if (total == right - left + 1) {
                res.add(s.substring(left, right + 1));
                queue.clear();
            }
        }

        return res;
    }
}

File: 1520-maximum-number-of-non-overlapping-substrings/1520-maximum-number-of-non-overlapping-substrings.cpp
class Solution {
public:
    vector<string> maxNumOfSubstrings(string s) {
        int count[26] = {};
        int first[26], last[26];

        fill(first, first + 26, -1);
        fill(last, last + 26, -1);

        vector<int> order;

        for (int i = 0; i < s.size(); i++) {
            int c = s[i] - 'a';

            if (count[c] == 0) {
                first[c] = i;
                order.push_back(c);
            }

            count[c]++;
            last[c] = i;
        }

        vector<string> res;
        deque<array<int, 3>> queue;

        for (int c : order) {
            queue.push_front({first[c], last[c], count[c]});

            int left = INT_MAX;
            int right = INT_MIN;
            int total = 0;

            for (auto& item : queue) {
                total += item[2];
                left = min(left, item[0]);
                right = max(right, item[1]);

                if (total == right - left + 1) {
                    break;
                }
            }

            if (total == right - left + 1) {
                res.push_back(s.substr(left, right - left + 1));
                queue.clear();
            }
        }

        return res;
    }
};